src/server.cpp

```diff
@@ -127,6 +127,98 @@ void *ServerThread::run()
 	return nullptr;
 }
 
+class MapThread : public Thread
+{
+public:
+
+	MapThread(Server *server):
+		Thread("Server"),
+		m_server(server)
+	{}
+
+	void *run();
+
+private:
+	Server *m_server;
+};
+
+void *MapThread::run()
+{
+	BEGIN_DEBUG_EXCEPTION_HANDLER
+
+	static thread_local const float save_interval =
+		g_settings->getFloat("server_map_save_interval") * 1000;
+
+	u64 last_time = 0;
+	int flush_counter = 0;
+	while (!stopRequested()) {
+		sleep_ms(100); // no busy loop
+
+		u64 curr_time = porting::getTimeMs();
+		if (curr_time - last_time < save_interval)
+			continue;
+
+		last_time = curr_time;
+
+		ServerEnvironment &env = m_server->getEnv();
+		ServerMap &map = (ServerMap&) env.getMap();
+
+		ModifiedState state;
+		if (++flush_counter < 4) {
+			infostream << "Saving Map" << std::endl;
+			state = MOD_STATE_WRITE_NEEDED;
+		} else {
+			infostream << "Flushing Map" << std::endl;
+			// pre-flush some of these blocks
+			state = MOD_STATE_WRITE_AT_UNLOAD;
+			flush_counter = 0;
+		}
+
+		// Save map, players and auth stuff
+		CompressedBlockList compressed_blocks;
+		SerializedBlockList serialized_blocks;
+		{
+			MutexAutoLock lock(m_server->m_env_mutex);
+
+			ScopeProfiler sp(g_profiler, "Server: map saving (sum)");
+
+			// Save ban file
+			if (m_server->getBanManager().isModified()) {
+				m_server->getBanManager().save();
+			}
+
+			// save the map's metadata
+			map.saveMetadata();
+
+			// 1. Map: Collect all changed blocks
+			map.collectAndMark(state, serialized_blocks);
+
+			// Save players
+			env.saveLoadedPlayers();
+
+			// Save environment metadata
+			env.saveMeta();
+		}
+
+		infostream << "Collect " << porting::getTimeMs() - curr_time << std::endl;
+
+		// 2. Map: Compress blocks outside of env lock
+		map.compress(serialized_blocks, compressed_blocks);
+		infostream << "Compress " << porting::getTimeMs() - curr_time << std::endl;
+
+		{
+			// 3. Map: Save blocks under lock
+			MutexAutoLock lock(m_server->m_env_mutex);
+			map.save(compressed_blocks);
```

Comment:
You can argue about the style here.
The other option is to pass the mutex to the existing `map::save(...)` method and handle it all there.